enzyme/Enzyme/MLIR/Implementations/FuncAutoDiffOpInterfaceImpl.cpp

```diff
@@ -29,9 +29,140 @@ namespace {
 #include "Implementations/FuncDerivatives.inc"
 } // namespace
 
+static std::optional<mlir::FunctionOpInterface>
+getContainingFunction(Operation *orig) {
+  Operation *parent;
+  while (parent = orig->getParentOp()) {
+    if (auto func = dyn_cast<mlir::FunctionOpInterface>(parent)) {
+      return std::optional(func);
+    }
+  }
+
+  return std::nullopt;
+}
+
+class AutoDiffCallRev
+    : public ReverseAutoDiffOpInterface::ExternalModel<AutoDiffCallRev,
+                                                       func::CallOp> {
+public:
+  LogicalResult createReverseModeAdjoint(Operation *orig, OpBuilder &builder,
+                                         MGradientUtilsReverse *gutils,
+                                         SmallVector<Value> caches) const {
+    DerivativeMode mode = DerivativeMode::ReverseModeGradient;
+
+    SymbolTable symbolTable = SymbolTable::getNearestSymbolTable(orig);
+
+    func::CallOp callOp = cast<func::CallOp>(orig);
+
+    auto parent = getContainingFunction(orig);
+    if (parent.has_value() &&
+        callOp.getCallee() == parent.value().getNameAttr()) {
```

Comment:
Ah I see this use, this is super fragile. For example it won’t work if you have a calls b calls a.

I think this is okay to merge as is but adding proper recursion support shouldn’t be bad. We have it already for forward enzymemlir which basically just requires having a cache in enzymelogic asking if we’ve already made this derivative function

Reply:
yeah any cycle in the call graph will not be caught here. also it would be nice if we could somehow store the symbol corresponding to the derivative of that call (with corresponding activities) to call the same reverse function in subsequent calls.

Reply:
Yeah enzymelogic should do this for forwards (and also llvm) already

Reply:
I will look at implementing something like ForwardCachedFunctions for reverse in a follow-up